libass/ass_render.c

```diff
@@ -990,25 +1076,122 @@ static void fill_composite_hash(CompositeHashKey *hk, CombinedBitmapInfo *info)
+        if (info->font->desc.vertical && info->symbol >= VERTICAL_LOWER_BOUND)
+            k->flags |= DECO_ROTATE;
+
+        val = ass_cache_get(priv->cache.outline_cache, &key, priv);
+        if (!val || !val->valid) {
+            ass_cache_dec_ref(val);
+            return;
+        }
+
+        scale.x = info->scale_x;
+        scale.y = info->scale_y;
+        asc  = val->asc;
+        desc = val->desc;
+    }
+
+    info->outline = val;
+    info->outline_transform.scale = scale;
+    info->outline_transform.offset = offset;
+
+    info->bbox.x_min = lrint(val->cbox.x_min * scale.x + offset.x);
+    info->bbox.y_min = lrint(val->cbox.y_min * scale.y + offset.y);
+    info->bbox.x_max = lrint(val->cbox.x_max * scale.x + offset.x);
+    info->bbox.y_max = lrint(val->cbox.y_max * scale.y + offset.y);
 
-    info->hash_key.u.outline.outline = val;
-    info->outline = &val->outline;
-    info->border[0] = &val->border[0];
-    info->border[1] = &val->border[1];
-    info->bbox = val->bbox_scaled;
     if (info->drawing_text || priv->settings.shaper == ASS_SHAPING_SIMPLE) {
-        info->cluster_advance.x = info->advance.x = val->advance.x;
-        info->cluster_advance.y = info->advance.y = val->advance.y;
+        info->cluster_advance.x = info->advance.x = lrint(val->advance * scale.x);
+        info->cluster_advance.y = info->advance.y = 0;
+    }
+    info->asc  = lrint(asc  * scale.y);
+    info->desc = lrint(desc * scale.y);
+
+    if (info->border_style == 3) {
+        key.type = OUTLINE_BOX;
+        val = ass_cache_get(priv->cache.outline_cache, &key, priv);
+        if (!val || !val->valid) {
+            ass_cache_dec_ref(val);
+            return;
+        }
+
+        double w = 64 * priv->border_scale;
+        ASS_DVector bord = { info->border_x * w, info->border_y * w };
+        double width = info->hspacing_scaled + info->advance.x;
+        double height = info->asc + info->desc;
+
+        ASS_Vector orig_scale;
+        orig_scale.x = info->scale_x * info->scale_fix;
+        orig_scale.y = info->scale_y * info->scale_fix;
+
+        // Emulate the WTFish behavior of VSFilter, i.e. double-scale
+        // the sizes of the opaque box.
+        bord.x *= orig_scale.x;
+        bord.y *= orig_scale.y;
+        width  *= orig_scale.x;
+        height *= orig_scale.y;
+
+        bord.x = FFMAX(64, bord.x);
+        bord.y = FFMAX(64, bord.y);
+
+        scale.x = (width  + 2 * bord.x) / 64;
+        scale.y = (height + 2 * bord.y) / 64;
+        offset.x = -bord.x;
+        offset.y = -bord.y - info->asc;
+    } else {
+        key.type = OUTLINE_BORDER;
+        BorderHashKey *k = &key.u.border;
+        k->outline = val;
+
+        double w = priv->border_scale * (64 / STROKER_ACCURACY);
+        ASS_DVector bord = { info->border_x * w, info->border_y * w };
+        k->border.x = lrint(bord.x / frexp(scale.x, &k->scale_ord_x));
+        k->border.y = lrint(bord.y / frexp(scale.y, &k->scale_ord_y));
```

Comment:
Yes for the first two questions.

About cache idea, I think, additional complexity of cache logic isn't worth performance gains, but it's good to consider that in the future. Especially if we'd have bitmap resources with similar logic (glyph distance fields, etc.).

As for parameters aside from `scale`, it's indeed wrong to disregard them. So technically that's a regression and two following commits address it. Should I mention that in commit message or add TODO comment in code?

Reply:
I don’t think it’s actually a regression, as we’ve always called the shaper before applying the transform, so the shaper could see only the scale (which we apply immediately) but not the other transform parameters.

Ah, I see, so that’s what those commits do. Sounds great.

If you agree that this is not a regression, then I don’t think any mention is required either in the commit message or in the code, but feel free to add one if you want (right now I’m slightly inclined towards a TODO comment, but it doesn’t matter much which you choose).

Reply:
It's real regression as I've especially constructed animation that looked jumpy due to the incorrect border quantization after that commit. Okay, I'll add TODO commit here.

Reply:
But we have never told the stroker about anything other than scale, have we? If this is a serious regression, then the only reason I can imagine is precisely this attempt to quantize the scale, not the fact that this doesn’t depend on any other parameters. In that case, should the quantization of scale perhaps be moved into the later commit where you add all the other parameters as well? Although I do admit that it’s not terribly useful when we’re pushing the whole PR at once.

Reply:
It's not that serious, in most cases that approximation is good enough. Only for large perspective rotations and possibly extreme skews it works unsatisfactory. Also some quantization is necessary anyway and I think that this approximation is better than arbitrary `double_to_d6()`.